### PR TITLE
🏹 Frontend serialization updates

### PR DESCRIPTION
**Pull request by lemald**
#### Summary of changes
**Asana Ticket:** [🏹 Frontend serialization updates](https://app.asana.com/0/584764604969369/1191956107867592/f)

Will need to coordinate merging and deploying with @losvedir's upcoming gtfs_creator PR.

#### Reviewer Checklist
- [ ] Meets ticket's acceptance criteria
- [ ] Any new or changed functions have typespecs
- [ ] Tests were added for any new functionality (don't just rely on Codecov)
- [ ] This branch was deployed to the staging environment and is currently running with no unexpected increase in warnings, and no errors or crashes.

### CODE REVIEW

File: assets/src/jsonApi.ts [hunk 1]
```diff
@@ -120,19 +124,33 @@ const loadRelationship = (
...
   return []
 }
 
+const loadSingleRelationship = (
+  relationship: any,
+  included: { [key: string]: ModelObject }
+): ModelObject | undefined => {
+  const loaded = loadRelationship(relationship, included)
+
+  if (loaded.length === 1) {
```
Comment (fjlanasa) [inline assets/src/jsonApi.ts:141]: again, can just access the first element in the array even if it's empty for the same effect

File: assets/src/models/disruption.ts [hunk 2]
```diff
@@ -51,29 +40,10 @@ class Disruption extends JsonApiResourceObject {
...
-        },
-        days_of_week: {
-          data: this.daysOfWeek.map((dow) => dow.toJsonApiData()),
-        },
-        exceptions: { data: this.exceptions.map((ex) => ex.toJsonApiData()) },
-        trip_short_names: {
-          data: this.tripShortNames.map((tsn) => tsn.toJsonApiData()),
-        },
-      },
+      attributes: "",
```
Comment (fjlanasa) [inline assets/src/models/disruption.ts:46]: i guess this doesn't really matter, but should we just set this to `{}` to make it consistent with the other functions?

File: lib/arrow_web/controllers/api/disruption_controller.ex [hunk 3]
```diff
@@ -4,50 +4,36 @@ defmodule ArrowWeb.API.DisruptionController do
...
-
-      :error ->
-        conn |> put_status(400) |> render(:errors, errors: [%{detail: "Invalid request"}])
-    end
+  def index(conn, _params) do
+    data =
+      from(d in Disruption,
+        join: dr in assoc(d, :revisions),
+        order_by: [d.id, dr.id],
+        where: dr.id >= d.published_revision_id or is_nil(d.published_revision_id),
```
Comment (fjlanasa) [inline lib/arrow_web/controllers/api/disruption_controller.ex:13]: this `is_nil(d.published_revision_id)` just returns all revisions if the disruption has no published revision?

File: assets/tests/disruptions/viewDisruption.test.tsx [hunk 4]
```diff
@@ -23,33 +24,72 @@ describe("ViewDisruption", () => {
       return Promise.resolve(
         new Disruption({
           id: "1",
-          startDate: new Date("2020-01-15"),
-          endDate: new Date("2020-01-30"),
-          adjustments: [
-            new Adjustment({
+          readyRevision: new DisruptionRevision({
```
Comment (fjlanasa) [inline assets/tests/disruptions/viewDisruption.test.tsx:27]: might want to think about looking into some sort of factory library for these in a follow up ticket

**Review by fjlanasa — APPROVED** [review]
whew, so many `disruption` -> `disruptionRevision`. a few small questions but overall looks good!

File: assets/src/disruptions/viewDisruption.tsx [hunk 5]
```diff
@@ -95,25 +97,24 @@ interface ViewDisruptionFormProps {
...
-        encodeURIComponent(disruptionId) +
-        `?only_ready=${view === DisruptionView.Ready ? "true" : "false"}`,
+      url: "/api/disruptions/" + encodeURIComponent(disruptionId),
       parser: toModelObject,
       defaultResult: "error",
     }).then((result: JsonApiResponse) => {
       if (result instanceof Disruption) {
-        setDisruption(result)
+        setDisruptionRevision(
+          revisionFromDisruptionForView(result, view) as DisruptionRevision
```
[the remark this answers is not in the document]
Comment (lemald) [reply assets/src/disruptions/viewDisruption.tsx:114]: Ah it looks like I can do `|| null` instead to handle the case where the revisions is undefined, which I like a lot more.

File: assets/src/disruptions/viewDisruption.tsx [hunk 6]
```diff
@@ -147,27 +152,32 @@ const ViewDisruptionForm = ({
...
-                tripShortNames={disruption.tripShortNames
+                tripShortNames={disruptionRevision.tripShortNames
                   .map((tsn) => tsn.tripShortName)
                   .join(", ")}
               />
               {view === DisruptionView.Draft && (
                 <>
                   <div>
-                    <EditDisruptionButton disruptionId={disruption.id} />
+                    {disruptionRevision.disruptionId && (
```
[the remark this answers is not in the document]
Comment (lemald) [reply assets/src/disruptions/viewDisruption.tsx:168]: Hmm, how would that work with serializing a newly-created disruptionRevision, though, which hasn't yet been saved to the database? That's the main reason I didn't make it required.

File: lib/arrow_web/controllers/api/disruption_controller.ex (hunk 3 above)
Comment (lemald) [reply lib/arrow_web/controllers/api/disruption_controller.ex:13]: Correct - the goal is for the logic to basically return everything more recent than the published revision, inclusive, and if no revision is yet published then return everything.

File: assets/tests/disruptions/viewDisruption.test.tsx (hunk 4 above)
Comment (lemald) [reply assets/tests/disruptions/viewDisruption.test.tsx:27]: Agreed, fiddling with the test data while working on this ticket got really tedious...

File: assets/src/disruptions/viewDisruption.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (fjlanasa) [reply assets/src/disruptions/viewDisruption.tsx:168]: yeah, fair point. i was thinking about it in the context of loading an existing one, but not using it for a yet to be created one.

**Pull request merged**